language_service/src/state.rs

```diff
@@ -0,0 +1,465 @@
+                        // only report the errors for one of them, the goal being
+                        // a less confusing user experience.
+                        continue;
+                    }
+
+                    self.publish_diagnostics_for_doc(state, &uri, errors);
+                }
+            }
+
+            // Clear errors from any documents that previously had errors
+            for uri in last_docs_with_errors.difference(&docs_with_errors) {
+                self.publish_diagnostics_for_doc(state, uri, vec![]);
+            }
+        });
+
+        self.documents_with_errors = docs_with_errors;
+    }
+
+    fn publish_diagnostics_for_doc(&self, state: &CompilationState, uri: &str, errors: Vec<Error>) {
+        let version = state.open_documents.get(uri).map(|d| d.version);
+        trace!(
+            "publishing diagnostics for {uri} {version:?}): {} errors",
+            errors.len()
+        );
+        (self.diagnostics_receiver)(DiagnosticUpdate {
+            uri: uri.into(),
+            version,
+            errors,
+        });
+    }
+
+    fn apply_configuration(&mut self, configuration: &WorkspaceConfigurationUpdate) -> bool {
+        let mut need_recompile = false;
+
+        if let Some(package_type) = configuration.package_type {
+            need_recompile |= self.configuration.package_type != package_type;
+            self.configuration.package_type = package_type;
+        }
+
+        if let Some(target_profile) = configuration.target_profile {
+            need_recompile |= self.configuration.target_profile != target_profile;
+            self.configuration.target_profile = target_profile;
+        }
+
+        // Possible optimization: some projects will have overrides for these configurations,
+        // so workspace updates won't impact them. We could exclude those projects
+        // from recompilation, but we don't right now.
+        trace!("need_recompile after configuration update: {need_recompile}");
+        need_recompile
+    }
+
+    /// Recompiles the currently known documents with
+    /// the current configuration. Publishes updated
+    /// diagnostics for all documents.
+    fn recompile_all(&mut self) {
+        self.with_state_mut(|state| {
+            for compilation in state.compilations.values_mut() {
+                let configuration = merge_configurations(&compilation.1, &self.configuration);
+                compilation
+                    .0
+                    .recompile(configuration.package_type, configuration.target_profile);
+            }
+        });
+
+        self.publish_diagnostics();
+    }
+
+    fn with_state<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(&CompilationState) -> T,
+    {
+        let state = self.state.borrow();
+        f(&state)
+    }
+
+    fn with_state_mut<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(&mut CompilationState) -> T,
+    {
+        let mut state = self.state.borrow_mut();
```

Comment:
Yeah good concern, I'm absolutely certain this won't panic. I can explain in a comment too, but here's the breakdown:

So, the `CompilationStateUpdater` isn't `Send`/`Sync` (nothing is - we're always single threaded). So we have this blanket guarantee that this state will never be shared across threads. That's one reassurance.

So the only remaining thing to be concerned about would be `await` points. The current code doesn't keep the mutable borrow across an await point. And it never ever should in the future. There is a comment warning against this:
https://github.com/microsoft/qsharp/blob/19f5fb5cd6ec59cf8eaf7f4ac418f75866dd509c/language_service/src/state.rs#L88-L91

In case the comment doesn't have enough teeth there's also a Clippy warning (which saved me many times in development)
https://rust-lang.github.io/rust-clippy/master/#/await_holding_refcell_ref

I'll add another comment right by this `borrow_mut` so no one is ever tempted to change it to `async` in the future.

Reply:
I don't really know how to protect against your examples except to say, "don't do that"? I don't think the patterns you showed exist in the code today, and I agree they're bad. Your examples have nested borrows in the same function, which is fairly easy to catch in development. I tried to address the less obvious cases (await points) in my other comment. 

Anything else I can do to cover all bases here?

Reply:
I'll leave a big scary comment. Not disagreeing that care should be taken. 

In general I'd suggest keeping `with_borrow_mut` scopes really small so you don't accidentally introduce reentrancy. But I feel like you'd have to be trying pretty hard to shoot yourself in the foot.